### ISSUE TITLE
Improvements to CONTRIBUTING.md: General proofreading

### ISSUE THREAD

**Issue by jtoar** [redwoodjs/redwood#439]
This issue addresses the first task in #434 (Improvements to CONTRIBUTING.md):

> General proofreading: description of process and information -- does it make sense to a first-time reader and does it work when someone sets up from scratch

Since it's painful to edit markdown directly and virtually impossible to discuss edits without throwing up a PR (and even then...) we'll use a [google doc](https://docs.google.com/document/d/1wsUpLwniebKxMI_wgkwMJfljJzyIHzj21fhGrGsWdGg/edit?usp=sharing) to make and discuss edits. If all goes well, I'll write up this workflow as a stand-alone file.

### If we're using a google doc, what are comments in this Issue for? 

Via @thedavidprice (I'm more-or-less quoting here) use comments in this Issue for process/project management. I.e. “Hey, I did another round of edits and they’re ready for your review.” Do this to trigger the next step.

This is another way of saying: **don't accept/reject/resolve a suggestion if you made it**. Wait for feedback--ask for feedback.

### What's good editing?

As @thedavidprice mentioned in #434, the Create React App [CONTRIBUTING.md](https://github.com/facebook/create-react-app/blob/master/CONTRIBUTING.md) doc is a good reference. I'd recommend reading that first.

In terms of editing-at-large (I'm speaking only for myself here), I'm not a professional, but I like Garner. I'll refer to [this piece](https://www.abajournal.com/magazine/article/garner_conversational_writing) of his (don't dismiss it just because of its title). TL;DR:

> ...all good editors do three things to a writer’s style: tighten (by combating verbosity), sharpen (by combating vagueness) and brighten (by combating dullness). A fourth, error correction, isn’t so much a matter of style as preventing various types of outright blunders—everything from factual mistakes to solecisms.

### What's this issue for?

As aforementioned, general proofreading.

> Note: one of the criteria is that this works when someone sets up from scratch--`rwdev` doesn't actually work out of the box yet. Until v0.6.0, you'll have to make the symlink yourself.

**Comment by thedavidprice** [comment]
@jtoar Dom, I’ll add some notes for you and I throughout my questions below. Will just [Dom] to indicate. Here’s one to start with:
>[Dom] I like how Peter uses “Redwood framework” and “Redwood project” to differentiate between the repo redwoodjs/redwood and a Redwood App (from redwoodjs/create-redwood-app). I think it would be great to explain this up front and then use the RW Framework and RW Project nomenclature _clearly_ throughout to give people context. It’s confusing, unfortunately, by nature of the beast.

@peterp I’ve run through this locally. Questions I have below:
**Note**: I’m now using `rwt` in place of `rwdev`

### Watch and copy
I only need to use two commands for this process, correct?
- RW Framework: `yarn build:watch`
- RW Project: `yarn rwt watch`

>[Dom] it would be great to add to this section an example of creating an env var in `.env’. This should definitely be a sub header section.

### Emulating package publishing
This one feels a little harder to follow, I think because it’s easy to lose context and the code examples (which are visually highlighted) are _not_ the code to run.

But, to confirm, this is a three step process, correct? (Well, 3 steps _after_ you install verdaccio.)`
1. RW Framework: `./tasks/run-local-npm`
2. RW Framework: `./tasks/publish-local`
3. RW Project: `yarn rwt install [local-package-name]`

- Can I use this to install other npm packages from my Framework into my Project? E.g. when I was doing CLI work I had to add `cross-env`, `rimraf`, etc, which I also had to add to my Project in order to fully test. (Sometimes this also seems to be needed in the case of upgrading existing packages, yes?)  Are these new tools meant to help with any of this? Fine if not, but we definitely need to cover in Contributing Docs either way. ([Dom] please help me track this part.)
- Any thoughts/reactions to creating `package.json` scripts for the scripts in "tasks/“? It might be nice to have one script that starts server and publishes all at startup. If this feels helpful, I don’t know if we need cover all cases, e.g. install individual packages.
- There’s another `rwt` command `fix-bins  Fix Redwood symlinks and permissions`, when, if ever, would this need to be used?

>[Dom] Several things here. We need to do a better job with clarity and division between Setup, How-to Steps, and Going Deeper. We need to clarify how to install _any/all_ of the local `@redwoodjs` packages and address how to handle other types of package (see above)

>[Dom] Need to call out that you can either “build and publish all” by default, or you can just build and publish a _specific_ package. See this —> "You can build a particular package by specifying the path to the package: ./tasks/publish-local ./packages/api."

Thanks, all 🎉

**Comment by jtoar** [comment]
@thedavidprice 
> [Dom] I like how Peter uses “Redwood framework” and “Redwood project” to differentiate between the repo redwoodjs/redwood and a Redwood App (from redwoodjs/create-redwood-app).

Working on this now. What do you think of exchanging "Redwood Project" for "Redwood App" since the command's called yarn create redwood-app?

**Comment by jtoar** [comment]
@thedavidprice I've added two versions [V1] and [V2] to the **Local Development** section addressing the "Redwood Framework"-"Redwood Project" nomenclature. Let me know what you think so far. I'll continue to tune them/add more.

**Comment by thedavidprice** [comment]
I’m in favor of "Redwood App”. 

@peterp do you have a preference?

**Comment by jtoar** [comment]
@thedavidprice Added a subheader section with an example of creating a RW_PATH env var ([link to subheader section](https://docs.google.com/document/d/1wsUpLwniebKxMI_wgkwMJfljJzyIHzj21fhGrGsWdGg/edit#heading=h.klpvjp41az8l)--made this using a bookmark in google docs). I use Ubuntu, and I think the process is similar on a Mac (just replace .bashrc for .bash_profile), but do you know how to document this for Windows? I think you could do it many ways?

**Comment by thedavidprice** [comment]
Ah, I hadn’t specifically read the code to understand where the env var was located. My assumption was that it was stored in the App directory `.env` file and using dotenv to accesss. Which I would prefer rather than store on system.

Another option is to use the `redwoodjs.toml` like this:
```javascript
import { getConfig } from '@redwoodjs/internal’

...

getConfig().redwood-tools.RW_PATH
```

In `redwood.toml`, we’d be looking for:
```javascript
[redwood-tools]
  RW_PATH = "I’m not sure if this should be relative or absolute path to Project"
```

@jtoar and @peterp Reactions to either of these implementation options above?

**Comment by peterp** [comment]
> My assumption was that it was stored in the App directory .env file and using dotenv to accesss. Which I would prefer rather than store on system.

The redwood tool CLI is used for framework developers - the assumption is that they would have a single place where they checked out the redwood framework, and adding this to a single project would be OK, but I don't think that it's any easier or better than all the other options we have:

There are 2 ways to do this with an env var: 
1. Add it to `.bash_profile`
2. Prepend it to the start of the command.

Or, supply it as an argument to the command; `yarn rwt cpw ../relative/path/to/redwood-framework`

**Comment by peterp** [comment]
I guess I want to understand why you would prefer it in the redwood.toml file?

**Comment by jtoar** [comment]
@thedavidprice The vars in `.env` seems like they're more for production? Not to say the file couldn't take on more of a role. I like the `.toml` example you gave--that looks clean. The reason I wrote it up the way I did is I'm used to an env var === bash so I just assumed that was what @peterp meant.

**Comment by thedavidprice** [comment]
I like this question. And it was a response to the mental model I was imagining for myself as a user _combined_ with seeing it as a relative path in the example `yarn rwt cpw ../relative/path/to/redwood-framework`

I guess you could say my reasoning was based on an App-centric workflow (and not a Framework-centric workflow). Along with a specific repo I use for testing, I’m often spinning up Apps for one-off implementation — these are normally in the same directory and relative location to my Framework, but not always. So, again, I was likely stuck thinking relative-path-to-framework.

Having now walked through my own thinking, I retract my suggestion. Let’s stick with the system ENV var approach with instructions for adding to `.bash_profile`. Thanks!

> @peterp when you have the time, please do see my comment with questions for you, mostly regarding confirmation of understanding, which will help documentation clarity.

> @jtoar this conversation’s elephant-in-the-room is Windows specific 1) support and 2) documentation. It’s unlikely these new tools are going to work with Powershell. I think I’ll ask Sam Weaver to help us with local testing and then we can make some decisions based on those results. Make sense and sound like a plan?

**Comment by jtoar** [comment]
@thedavidprice Good point--I spin up one-off apps for testing too. If I had to specify the `RW_PATH` in the one-off's `.env` or `.toml`, it'd be an extra step. You get app-to-app permanence when its in bash. 

And bringing Sam Weaver in sounds great!

**Comment by thedavidprice** [comment]
@jtoar I just went through the Google Doc. Mostly added side comments but also contributed some edits. In general, we need to make sure the command instructions are correct before we release RedwoodJS v0.6.0 (hopefully tomorrow). But overall I think things are in great shape for a first round revision.

Could you take a look at my comments, tidy up the edits you wanted to make, and add this to #434 for final review to merge? 

Thanks again!

**Comment by jtoar** [comment]
@thedavidprice Yes; on it now.

**Comment by thedavidprice** [comment]
closed via #435
